Purchase extra storage from a modal

dsas opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Fixes DOTCOM-14998

## Proposed Changes

In the new Hosting Dashboard, use a modal to display the add-on storage options rather than linking to WordPress.com/add-ons/:siteSlug


## Why are these changes being made?
<!--
It's easy to see what a PR does but much harder to find out why it was made,
particularly when researching old changes in history. Record an explanation of
the motivation behind this change and how it will help.
-->

It keeps the user in the hosting dashboard, rather than pushing them out to wp-admin. This makes sense as it's a hosting concern rather than a wp-admin one.

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
/plans | New modal
-------|------
 <img width="562" height="302" alt="Screenshot 2025-10-21 at 16 28 11" src="https://github.com/user-attachments/assets/c58b602f-0373-4f02-8a91-ba79f25a244a" />  | <img width="562" height="428" alt="Screenshot 2025-10-21 at 20 01 46" src="https://github.com/user-attachments/assets/be8b3649-4554-426b-9098-aa55e86fb4c8" />

For the button that launches the modal to be displayed you need to be close or over the storage capacity limit. The easiest way to achieve this is to checkout the branch locally, update [storage-warning-banner.tsx](https://github.com/Automattic/wp-calypso/blob/79213813451e7ffbf68ea90b48858a199538d589/client/dashboard/sites/overview/storage-warning-banner.tsx#L26) and [site-storage-stat.tsx](https://github.com/Automattic/wp-calypso/blob/79213813451e7ffbf68ea90b48858a199538d589/client/dashboard/sites/overview-plan-card/site-storage-stat.tsx#L31) to hardcode `alertLevel` to `warning` and then build as normal.

Another way of testing this is to update `Media_Storage::get_space_info` on your sandbox to hardcode the desired numbers for your test blog.

1. Go to /v2/sites/ and choose a site
2. Using the notice press 'Add more storage' to launch the modal
3. Try different amounts of storage, the labels / meter beneath should react
4. Press checkout, you should go to the checkout with the thing you've chosen


You should cross-reference with the functionality on https://wordpress.com/add-ons/:siteSlug - it's displayed very slightly differently, but ultimately it's another way of purchasing the same thing.

You should also see an 'Add more storage' link on the plan card which launches the same modal.

https://github.com/user-attachments/assets/09c27873-286a-4d31-ba3a-488043f73d17



## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

dsas committed:
Purchase extra storage from a modal

In the new Hosting Dashboard, purchasing extra storage did link to the
WordPress.com/add-ons/:siteSlug page, now it shows a modal before
sending the user to the checkout.

dsas committed:
Show storage tiers as incremental additions rather than as selecting a given tier

dsas committed:
Maybe clarify the comment a bit more

dsas committed:
Simplify the selected value

dsas committed:
Remove minimum_price_monthly_display - it went unused

xavier-lc commented on client/dashboard/sites/overview/storage-warning-banner.tsx:
```diff
@@ -49,29 +52,38 @@ export function StorageWarningBanner( { site }: { site: Site } ) {
+					}
+				) }
+			</Notice>
+			{ isModalOpen && (
```
Given that the modal already has `if ( ! isOpen ) {		return null;	}`, we might omit this so the queries get running in the meantime?

xavier-lc reviewed (approved):
Works as described 👌

fushar commented on client/dashboard/sites/storage/add-storage-modal.tsx:
```diff
@@ -0,0 +1,155 @@
+
+export function AddStorageModal( { site, isOpen, onClose }: AddStorageModalProps ) {
+	const { data: products } = useSuspenseQuery( productsQuery() );
+	const { data: purchases } = useSuspenseQuery( sitePurchasesQuery( site.ID ) );
```
Is this preloaded query ensured anywhere in the router?

dsas replied:
They're not preloaded, and I think it would be wasteful to do so. I'd assume that most people aren't near their storage limit and so won't see the "Add more storage" links at all, let alone click them.

Making the change @xavier-lc suggested to not wrap `AddStorageModal` in an `isModalOpen` check will alleviate this somewhat, but probably I guess it ought to use `useQuery` and display a spinner or something while loading.

What do you think?

dsas replied:
Gone ahead and done that in [1dd2402](https://github.com/Automattic/wp-calypso/pull/106621/commits/1dd24022a0a8e967803f3e290652c212a03a97ac)

fushar replied:
Hmm, in the current way, we'll ALWAYS fetch the products and purchases queries, even though the storage limit is not reached 🤔 

I think we should still use the `isModalOpen` check. Then, (only) when it's true, the queries will be fired and we show loading on the modal when they are not finished yet. What do you think?

fushar replied:
Oh! Maybe a better way is to put the modal into the `alertLevel !== 'none'` check (along with the CTA) so that the queries start whenever the site is low on storage.

dsas committed:
Remove wrapping of storage modal, to give the APIs requests more time to preload

dsas committed:
Change loading approach to show a spinner if the queries to get the data are not yet complete

fushar reviewed (commented):
It's a bit weird that these figures do not match. Should we use the same rounding logic here @lucasmendes-design (and if so, which one?)

<img width="863" height="387" alt="Image" src="https://github.com/user-attachments/assets/ac1041b1-882e-4052-ae6e-84111735bb10" />

dsas committed:
Align capacity rounding with overview

dsas committed:
Change loading of AddStorageModal

Only render the AddStorageModal component when nearing the capacity
limit.

The AddStorageModal component won't actually be rendered just because
its nearing the capacity limit due to its internal logic, but it will
start fetching the data needed to display just in case.

fushar reviewed (approved):
Looking good!
